### PR TITLE
#23267 Warn when Orderby is silently ignored when using Distinct

### PR DESCRIPTION
**Pull request by yesmey**
Implement a warning for when using OrderBy before Distinct. The OrderBy will silently get ignored or cause unintentional behavior.

The reason and design behind this change can be read in #23267

In simple terms, the definition of Enumerable.Distinct() is that it returns an unordered result...
> The Distinct<TSource>(IEnumerable<TSource>) method returns an unordered sequence that contains no duplicate values

... so when using a Distinct after a OrderBy - the previously specified order will get ignored

### CODE REVIEW

File: test/EFCore.Specification.Tests/Query/InheritanceQueryTestBase.cs [hunk 1]
```diff
@@ -457,9 +457,9 @@ public virtual Task Subquery_OfType(bool async)
             return AssertQuery(
                 async,
                 ss => ss.Set<Bird>()
+                    .Distinct()
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/Query/InheritanceQueryTestBase.cs:460]: This is a valid query. Distinct will only remove ordering if there is no skip/take afterwards. Else Ordering will be used to apply limit/offset. Please revert.

File: test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs [hunk 2]
```diff
@@ -4410,7 +4410,7 @@ public virtual Task Correlated_collections_with_Distinct(bool async)
         {
             return AssertQuery(
                 async,
-                ss => ss.Set<Squad>().OrderBy(s => s.Name).Select(s => s.Members.OrderBy(m => m.Nickname).Distinct()),
+                ss => ss.Set<Squad>().OrderBy(s => s.Name).Select(s => s.Members.Distinct().OrderBy(m => m.Nickname).ToList()),
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs:4413]: Please revert this change and apply Skip(0) after OrderBy to continue generating LeftJoin.

File: test/EFCore.Cosmos.FunctionalTests/Query/InheritanceQueryCosmosTest.cs [hunk 3]
```diff
@@ -358,18 +358,12 @@ public override async Task OfType_Union_OfType(bool async)
             AssertSql(" ");
         }
 
+        [ConditionalTheory(Skip = "Distinct ordering #16156")]
```
Comment (smitpatel) [inline test/EFCore.Cosmos.FunctionalTests/Query/InheritanceQueryCosmosTest.cs:361]: Why this test required to be skipped now?

**Comment by smitpatel** [comment]
@yesmey - Please hold on making any changes to PR. This doesn't feel exactly the warning we want to show. We will discuss with team what exactly we want to implement. Please monitor the tracking issue for any updates.

**Comment by smitpatel** [comment]
@yesmey - Please refer to design meeting notes on original issue and update the PR accordingly.

File: test/EFCore.Specification.Tests/Query/InheritanceQueryTestBase.cs (hunk 1 above)
Comment (yesmey) [reply test/EFCore.Specification.Tests/Query/InheritanceQueryTestBase.cs:460]: I will revert this and all my changes related to this test

File: test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs (hunk 2 above)
Comment (yesmey) [reply test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs]: It still generates a OUTER APPLY after the change. Any idea how to solve it it?

File: src/EFCore/Properties/CoreStrings.resx [hunk 4]
```diff
@@ -765,6 +765,10 @@
     <value>The query uses the 'First'/'FirstOrDefault' operator without 'OrderBy' and filter operators. This may lead to unpredictable results.</value>
     <comment>Warning CoreEventId.FirstWithoutOrderByAndFilterWarning</comment>
   </data>
+  <data name="LogDistinctAfterOrderBy" xml:space="preserve">
```
Comment (smitpatel) [inline src/EFCore/Properties/CoreStrings.resx:768]: @dotnet/efteam - For exception message review

File: src/EFCore/Diagnostics/CoreEventId.cs [hunk 5]
```diff
@@ -274,6 +275,17 @@ private static EventId MakeQueryId(Id id)
...
 
+        /// <summary>
+        ///     <para>
+        ///         A query uses the 'Distinct' operator after 'OrderBy'. The 'OrderBy' operator will be ignored and may lead to unpredictable results.
+        ///     </para>
+        ///     <para>
+        ///         This event is in the <see cref="DbLoggerCategory.Query" /> category.
+        ///     </para>
+        /// </summary>
+        public static readonly EventId DistinctAfterOrderByWarning
```
Comment (smitpatel) [inline src/EFCore/Diagnostics/CoreEventId.cs:286]: `DistinctAfterOrderByWithoutRowLimitingOperatorWarning

File: src/EFCore/Properties/CoreStrings.resx [hunk 6]
```diff
@@ -765,6 +765,10 @@
...
+    <value>The query uses the 'Distinct' operator after a 'OrderBy' operator. This may lead to unpredictable results.</value>
```
Comment (smitpatel) [inline src/EFCore/Properties/CoreStrings.resx:769]:
```suggestion
    <value>The query uses the 'Distinct' operator after applying an ordering. If there are any row limiting operation used before `Distinct` and after ordering then ordering will be used for it. Ordering(s) will be erased after `Distinct` and results afterwards would be unordered.</value>
```

File: src/EFCore/Properties/CoreStrings.resx [hunk 7]
```diff
@@ -910,7 +914,7 @@
     <comment>Debug CoreEventId.RequiredAttributeOnSkipNavigation string string</comment>
   </data>
   <data name="LogRowLimitingOperationWithoutOrderBy" xml:space="preserve">
-    <value>The query uses a row limiting operator ('Skip'/'Take') without an 'OrderBy' operator. This may lead to unpredictable results.</value>
+    <value>The query uses a row limiting operator ('Skip'/'Take') without an 'OrderBy' operator. This may lead to unpredictable results. If the 'Distinct' operator is used, then make sure to use the 'OrderBy' operator after 'Distinct' as the order would otherwise get erased.</value>
```
Comment (smitpatel) [inline src/EFCore/Properties/CoreStrings.resx:917]:
```suggestion
    <value>The query uses a row limiting operator ('Skip'/'Take') without an 'OrderBy' operator. This may lead to unpredictable results. If the 'Distinct' operator is used after `OrderBy`, then make sure to use the 'OrderBy' operator after 'Distinct' as the ordering would otherwise get erased.</value>
```

**Comment by yesmey** [comment]
@smitpatel please review the changes in [6e92cdaf779e809954efb43c5de085d1d9bab841](https://github.com/dotnet/efcore/pull/24160/commits/6e92cdaf779e809954efb43c5de085d1d9bab841)

**Review by smitpatel — APPROVED** [review]

**Comment by smitpatel** [comment]
@yesmey - Thank you for contribution.
